File: src/Application Framework/System/Configuration.vb
Namespace Entity

    Namespace Sys

        Public Class Configuration

            Public Sub New()
                Dim ConfigurationDetails As New DataSet

                Dim configRow As Integer = 0

                'Select config file for release type
                If IsRFT Then
                    ConfigurationDetails.ReadXml(Application.StartupPath & "\ApplicationSettings_RFT.config")
                ElseIf IsRFTTEST Then
                    ConfigurationDetails.ReadXml(Application.StartupPath & "\ApplicationSettings_RFTTEST.config")
                ElseIf IsBlueflame Then
                    ConfigurationDetails.ReadXml(Application.StartupPath & "\ApplicationSettings_Blueflame.config")
                ElseIf IsBlueflameTest Then
                    ConfigurationDetails.ReadXml(Application.StartupPath & "\ApplicationSettings_BlueflameTest.config")
                ElseIf IsGasway Or IsRelease Then
                    ConfigurationDetails.ReadXml(Application.StartupPath & "\ApplicationSettings.config")
                Else
                    'if were in beta, we have access to all configs
                    ConfigurationDetails.ReadXml(Application.StartupPath & "\ApplicationSettings_Beta.config")

                    Dim dbs As New DataTable
                    dbs.Columns.Add("ValueMember")
                    dbs.Columns.Add(New DataColumn("DisplayMember"))
                    dbs.Columns.Add(New DataColumn("Deleted"))

                    Dim index As Integer = 1
                    For Each config As DataRow In ConfigurationDetails.Tables(0).Rows
                        'if we can connect to db then add it to selection
                        If TestConnection(config.Item("DatabaseServer"), config.Item("DatabaseName"),
                                          config.Item("DatabaseUsername"), config.Item("DatabasePassword")) Then

                            dbs.Rows.Add(New String() {index, config.Item("DatabaseFriendlyName"), "0"})
                        End If
                        index += 1
                    Next

                    'if we have more than one connection then we default display selection window
                    If dbs.Rows.Count > 1 Then
                        Dim f As New FRMSelectDatabase(dbs)
                        f.ShowDialog()
                        configRow = f.SelectedDb
                    End If

                End If

                'set system configuration properties
                Dim value As String = ConfigurationDetails.Tables(0).Rows(configRow).Item("DatabaseName")
                Try
                    DBName = [Enum].Parse(GetType(Enums.DataBaseName), value)
                Catch ex As Exception
                    MsgBox("Enum for database missing", MessageBoxButtons.OK, MessageBoxIcon.Error)
                    App.CloseApplication()
                End Try

                DatabaseServer = ConfigurationDetails.Tables(0).Rows(configRow).Item("DatabaseServer")
                DatabaseUsername = ConfigurationDetails.Tables(0).Rows(configRow).Item("DatabaseUsername")
                DatabasePassword = ConfigurationDetails.Tables(0).Rows(configRow).Item("DatabasePassword")
                DatabaseName = ConfigurationDetails.Tables(0).Rows(configRow).Item("DatabaseName")
                DocumentsLocation = ConfigurationDetails.Tables(0).Rows(configRow).Item("DocumentsLocation")
                DatabaseFriendlyName = ConfigurationDetails.Tables(0).Rows(configRow).Item("DatabaseFriendlyName")
                TemplateLocation = ConfigurationDetails.Tables(0).Rows(configRow).Item("TemplateLocation")

                DatabasePort = ConfigurationDetails.Tables(0).Rows(configRow).Item("DatabasePort")
                SuperAdminID = ConfigurationDetails.Tables(0).Rows(configRow).Item("SuperAdminID")
                SupportEmailFrom = ConfigurationDetails.Tables(0).Rows(configRow).Item("SupportEmailFrom")
                SupportEmailTo = ConfigurationDetails.Tables(0).Rows(configRow).Item("SupportEmailTo")
                SalesEmailFrom = ConfigurationDetails.Tables(0).Rows(configRow).Item("SalesEmailFrom")
                InfoEmail = ConfigurationDetails.Tables(0).Rows(configRow).Item("FeedbackEmailFrom")
                SMTPServer = ConfigurationDetails.Tables(0).Rows(configRow).Item("SMTPServer")
                SMTPServerUsername = ConfigurationDetails.Tables(0).Rows(configRow).Item("SMTPServerUsername")
                SMTPServerPassword = ConfigurationDetails.Tables(0).Rows(configRow).Item("SMTPServerPassword")
                VATNO = ConfigurationDetails.Tables(0).Rows(configRow).Item("VATNO")

                CompanyName = ConfigurationDetails.Tables(0).Rows(configRow).Item("CompanyName")
                CompanyAddres1 = ConfigurationDetails.Tables(0).Rows(configRow).Item("CompanyAddres1")
                CompanyAddres2 = ConfigurationDetails.Tables(0).Rows(configRow).Item("CompanyAddres2")
                CompanyAddres3 = ConfigurationDetails.Tables(0).Rows(configRow).Item("CompanyAddres3")
                CompanyAddres4 = ConfigurationDetails.Tables(0).Rows(configRow).Item("CompanyAddres4")
                CompanyAddres5 = ConfigurationDetails.Tables(0).Rows(configRow).Item("CompanyAddres5")
                CompanyPostcode = ConfigurationDetails.Tables(0).Rows(configRow).Item("CompanyPostcode")
                CompanyTelephoneNumber = ConfigurationDetails.Tables(0).Rows(configRow).Item("CompanyTelephoneNumber")
                CompanyWebsite = ConfigurationDetails.Tables(0).Rows(configRow).Item("CompanyWebsite")
                CompanyDomain = ConfigurationDetails.Tables(0).Rows(configRow).Item("CompanyDomain")

                CompanyFullAddress = CompanyName & ", " & CompanyAddres1 & ", " & CompanyAddres2 & ", " & CompanyAddres3 & ", " & CompanyPostcode
            End Sub

            Private Function TestConnection(ByVal dbServer As String, ByVal dbName As String,
                                            ByVal dbUsername As String, ByVal dbPass As String) As Boolean

                'set up connection string
                Dim str As String = ""
                If (dbServer.Contains(".\")) Then
                    str = "Data Source=" & dbServer & ";Initial Catalog=" & dbName & ";Integrated Security=True"
                Else
                    str = "Server=tcp:" & dbServer & ";"
                    str += "User Id=" & dbUsername & ";"
                    If dbPass.Trim.Length > 0 Then
                        str += "Password=" & dbPass & ";"
                    End If
                    str += "Database=" & dbName & ";"
                End If

                'TEST CONNECTION
                Try
                    Using conn As SqlClient.SqlConnection = New SqlClient.SqlConnection(str)
                        Try
                            conn.Open()
                            Return True
                        Catch ex As Exception
                        Finally
                            If conn.State = ConnectionState.Open Then
                                conn.Close()
                            End If
                        End Try
                    End Using
                Catch ex As Exception
                End Try
            End Function

#Region "Configuration"

            Private _isDebug As Boolean = False

            Public ReadOnly Property IsDebug As Boolean
                Get
#If DEBUG Then
                    _isDebug = True
#End If
                    Return _isDebug
                End Get
            End Property

            Private _isRelease As Boolean = False

            Public ReadOnly Property IsRelease As Boolean
                Get
#If RELEASE Then
                    _isRelease = True
#End If
                    Return _isRelease
                End Get
            End Property

            Private _isGasway As Boolean = False

            Public ReadOnly Property IsGasway As Boolean
                Get
#If GASWAY Then
                    _isGasway = True
#End If
                    Return _isGasway
                End Get
            End Property

            Private _isRFT As Boolean = False

            Public ReadOnly Property IsRFT As Boolean
                Get
#If RFT Then
                    _isRFT = True
#End If
                    Return _isRFT
                End Get
            End Property

            Private _isRFTTEST As Boolean = False

            Public ReadOnly Property IsRFTTEST As Boolean
                Get
#If RFTTEST Then
                    _isRFTTEST = True
#End If
                    Return _isRFTTEST
                End Get
            End Property

            Private _isBlueflame As Boolean = False

            Public ReadOnly Property IsBlueflame As Boolean
                Get
#If Blueflame Then
                    _isBlueflame = True
#End If
                    Return _isBlueflame
                End Get
            End Property

            Private _isBlueflameTest As Boolean = False

            Public ReadOnly Property IsBlueflameTest As Boolean
                Get
#If BlueflameTest Then
                    _isBlueflameTest = True
#End If
                    Return _isBlueflameTest
                End Get
            End Property

#End Region

#Region "Database"

            Private _DatabaseServer As String = String.Empty

            Public Property DatabaseServer() As String
                Get
                    Return _DatabaseServer
                End Get
                Set(ByVal Value As String)
                    _DatabaseServer = Value
                End Set
            End Property

            Private _DatabaseFriendlyName As String = String.Empty

            Public Property DatabaseFriendlyName() As String
                Get
                    Return _DatabaseFriendlyName
                End Get
                Set(ByVal Value As String)
                    _DatabaseFriendlyName = Value
                End Set
            End Property

            Private _DBName As Entity.Sys.Enums.DataBaseName

            Public Property DBName() As Entity.Sys.Enums.DataBaseName
                Get
                    Return _DBName
                End Get
                Set(ByVal value As Entity.Sys.Enums.DataBaseName)
                    _DBName = value
                End Set
            End Property

            Private _DatabaseUsername As String = String.Empty

            Public Property DatabaseUsername() As String
                Get
                    Return _DatabaseUsername
                End Get
                Set(ByVal Value As String)
                    _DatabaseUsername = Value
                End Set
            End Property

            Private _DatabasePassword As String = String.Empty

            Public Property DatabasePassword() As String
                Get
                    Return _DatabasePassword
                End Get
                Set(ByVal Value As String)
                    _DatabasePassword = Value
                End Set
            End Property

            Private _DatabaseName As String = String.Empty

            Public Property DatabaseName() As String
                Get
                    Return _DatabaseName
                End Get
                Set(ByVal Value As String)
                    _DatabaseName = Value
                End Set
            End Property

            Private _DatabasePort As String = String.Empty

            Public Property DatabasePort() As String
                Get
                    Return _DatabasePort
                End Get
                Set(ByVal Value As String)
                    _DatabasePort = Value
                End Set
            End Property

            Public ReadOnly Property ConnectionString() As String
                Get
                    Dim str As String = ""
                    Select Case TheSystem.DataBaseServerType
                        Case Enums.DatabaseSystems.MySQL
                            str += "server=" & DatabaseServer & ";"
                            str += "user id=" & DatabaseUsername & ";"
                            If DatabasePassword.Trim.Length > 0 Then
                                str += "password=" & DatabasePassword & ";"
                            End If
                            str += "database=" & DatabaseName & ";"
                            str += "port=" & DatabasePort & ";"
                        Case Enums.DatabaseSystems.Microsoft_SQL_Server
                            If (DatabaseServer.Contains(".\")) Then
                                str = "Data Source=" & DatabaseServer & ";Initial Catalog=" & DatabaseName & ";Integrated Security=True"
                            Else
                                str = "Server=tcp:" & DatabaseServer & ";"
                                str += "User Id=" & DatabaseUsername & ";"
                                If DatabasePassword.Trim.Length > 0 Then
                                    str += "Password=" & DatabasePassword & ";"
                                End If
                                str += "Database=" & DatabaseName & ";"
                            End If
                    End Select

                    Return str
                End Get
            End Property

#End Region

#Region "System Dependant"

            Public ReadOnly Property SystemVersion() As String
                Get
                    Return System.Diagnostics.FileVersionInfo.GetVersionInfo(System.Reflection.Assembly.GetExecutingAssembly.Location).FileVersion
                End Get
            End Property

            Private _SuperAdminID As Integer = 0

            Public Property SuperAdminID() As Integer
                Get
                    Return _SuperAdminID
                End Get
                Set(ByVal Value As Integer)
                    _SuperAdminID = Value
                End Set
            End Property

            Private _SupportEmailFrom As String = String.Empty

            Public Property SupportEmailFrom() As String
                Get
                    Return _SupportEmailFrom
                End Get
                Set(ByVal Value As String)
                    _SupportEmailFrom = Value
                End Set
            End Property

            Private _SupportEmailTo As String = String.Empty

            Public Property SupportEmailTo() As String
                Get
                    Return _SupportEmailTo
                End Get
                Set(ByVal Value As String)
                    _SupportEmailTo = Value
                End Set
            End Property

            Private _SalesEmailFrom As String = String.Empty

            Public Property SalesEmailFrom() As String
                Get
                    Return _SalesEmailFrom
                End Get
                Set(ByVal Value As String)
                    _SalesEmailFrom = Value
                End Set
            End Property

            Public Property InfoEmail() As String

            Private _SMTPServer As String = String.Empty

            Public Property SMTPServer() As String
                Get
                    Return _SMTPServer
                End Get
                Set(ByVal Value As String)
                    _SMTPServer = Value
                End Set
            End Property

            Private _SMTPServerUsername As String = String.Empty

            Public Property SMTPServerUsername() As String
                Get
                    Return _SMTPServerUsername
                End Get
                Set(ByVal Value As String)
                    _SMTPServerUsername = Value
                End Set
            End Property

            Private _SMTPServerPassword As String = String.Empty

            Public Property SMTPServerPassword() As String
                Get
                    Return _SMTPServerPassword
                End Get
                Set(ByVal Value As String)
                    _SMTPServerPassword = Value
                End Set
            End Property

            Private _DocumentsLocation As String = String.Empty

            Public Property DocumentsLocation() As String
                Get
                    Return _DocumentsLocation
                End Get
                Set(ByVal Value As String)
                    _DocumentsLocation = Value
                End Set
            End Property

            Private _TemplateLocation As String = String.Empty

            Public Property TemplateLocation() As String
                Get
                    Return _TemplateLocation
                End Get
                Set(ByVal Value As String)
                    _TemplateLocation = Value
                End Set
            End Property

            Private _VATNO As String = String.Empty

            Public Property VATNO() As String
                Get
                    Return _VATNO
                End Get
                Set(ByVal Value As String)
                    _VATNO = Value
                End Set
            End Property

            Public CompanyName As String
            Public CompanyAddres1 As String
            Public CompanyAddres2 As String
            Public CompanyAddres3 As String
            Public CompanyAddres4 As String
            Public CompanyAddres5 As String
            Public CompanyPostcode As String
            Public CompanyTelephoneNumber As String
            Public CompanyWebsite As String
            Public CompanyDomain As String
            Public CompanyFullAddress As String

#End Region

        End Class

    End Namespace

End Namespace

File: src/Application Framework/System/Configuration.cs
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using System;
using System.Data;
using System.Diagnostics;
using System.Windows.Forms;

namespace FSM.Entity
{
    namespace Sys
    {
        public class Configuration
        {
            public Configuration()
            {
                var ConfigurationDetails = new DataSet();
                int configRow = 0;

                // Select config file for release type
                if (IsRFT)
                {
                    ConfigurationDetails.ReadXml(Application.StartupPath + @"\ApplicationSettings_RFT.config");
                }
                else if (IsRFTTEST)
                {
                    ConfigurationDetails.ReadXml(Application.StartupPath + @"\ApplicationSettings_RFTTEST.config");
                }
                else if (IsBlueflame)
                {
                    ConfigurationDetails.ReadXml(Application.StartupPath + @"\ApplicationSettings_Blueflame.config");
                }
                else if (IsBlueflameTest)
                {
                    ConfigurationDetails.ReadXml(Application.StartupPath + @"\ApplicationSettings_BlueflameTest.config");
                }
                else if (IsGasway | IsRelease)
                {
                    ConfigurationDetails.ReadXml(Application.StartupPath + @"\ApplicationSettings.config");
                }
                else
                {
                    // if were in beta, we have access to all configs
                    ConfigurationDetails.ReadXml(Application.StartupPath + @"\ApplicationSettings_Beta.config");
                    var dbs = new DataTable();
                    dbs.Columns.Add("ValueMember");
                    dbs.Columns.Add(new DataColumn("DisplayMember"));
                    dbs.Columns.Add(new DataColumn("Deleted"));
                    int index = 1;
                    foreach (DataRow config in ConfigurationDetails.Tables[0].Rows)
                    {
                        // if we can connect to db then add it to selection
                        if (TestConnection(Conversions.ToString(config["DatabaseServer"]), Conversions.ToString(config["DatabaseName"]), Conversions.ToString(config["DatabaseUsername"]), Conversions.ToString(config["DatabasePassword"])))
                        {
                            dbs.Rows.Add(new string[] { index.ToString(), config["DatabaseFriendlyName"].ToString(), "0" });
                        }

                        index += 1;
                    }

                    // if we have more than one connection then we default display selection window
                    if (dbs.Rows.Count > 1)
                    {
                        var f = new FRMSelectDatabase(dbs);
                        f.ShowDialog();
                        configRow = f.SelectedDb;
                    }
                }

                // set system configuration properties
                string value = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["DatabaseName"]);
                try
                {
                    DBName = (Enums.DataBaseName)Enum.Parse(typeof(Enums.DataBaseName), value);
                }
                catch (Exception ex)
                {
                    Interaction.MsgBox("Enum for database missing", (MsgBoxStyle)MessageBoxButtons.OK, MessageBoxIcon.Error);
                    App.CloseApplication();
                }

                DatabaseServer = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["DatabaseServer"]);
                DatabaseUsername = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["DatabaseUsername"]);
                DatabasePassword = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["DatabasePassword"]);
                DatabaseName = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["DatabaseName"]);
                DocumentsLocation = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["DocumentsLocation"]);
                DatabaseFriendlyName = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["DatabaseFriendlyName"]);
                TemplateLocation = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["TemplateLocation"]);
                DatabasePort = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["DatabasePort"]);
                SuperAdminID = Conversions.ToInteger(ConfigurationDetails.Tables[0].Rows[configRow]["SuperAdminID"]);
                SupportEmailFrom = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["SupportEmailFrom"]);
                SupportEmailTo = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["SupportEmailTo"]);
                SalesEmailFrom = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["SalesEmailFrom"]);
                InfoEmail = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["FeedbackEmailFrom"]);
                SMTPServer = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["SMTPServer"]);
                SMTPServerUsername = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["SMTPServerUsername"]);
                SMTPServerPassword = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["SMTPServerPassword"]);
                VATNO = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["VATNO"]);
                CompanyName = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["CompanyName"]);
                CompanyAddres1 = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["CompanyAddres1"]);
                CompanyAddres2 = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["CompanyAddres2"]);
                CompanyAddres3 = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["CompanyAddres3"]);
                CompanyAddres4 = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["CompanyAddres4"]);
                CompanyAddres5 = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["CompanyAddres5"]);
                CompanyPostcode = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["CompanyPostcode"]);
                CompanyTelephoneNumber = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["CompanyTelephoneNumber"]);
                CompanyWebsite = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["CompanyWebsite"]);
                CompanyDomain = Conversions.ToString(ConfigurationDetails.Tables[0].Rows[configRow]["CompanyDomain"]);
                CompanyFullAddress = CompanyName + ", " + CompanyAddres1 + ", " + CompanyAddres2 + ", " + CompanyAddres3 + ", " + CompanyPostcode;
            }

            private bool TestConnection(string dbServer, string dbName, string dbUsername, string dbPass)
            {
                // set up connection string
                string str = "";
                if (dbServer.Contains(@".\"))
                {
                    str = "Data Source=" + dbServer + ";Initial Catalog=" + dbName + ";Integrated Security=True";
                }
                else
                {
                    str = "Server=tcp:" + dbServer + ";";
                    str += "User Id=" + dbUsername + ";";
                    if (dbPass.Trim().Length > 0)
                    {
                        str += "Password=" + dbPass + ";";
                    }

                    str += "Database=" + dbName + ";";
                }

                // TEST CONNECTION
                try
                {
                    using (var conn = new System.Data.SqlClient.SqlConnection(str))
                    {
                        try
                        {
                            conn.Open();
                            return true;
                        }
                        catch (Exception ex)
                        {
                        }
                        finally
                        {
                            if (conn.State == ConnectionState.Open)
                            {
                                conn.Close();
                            }
                        }
                    }
                }
                catch (Exception ex)
                {
                }

                return default;
            }

            /* TODO ERROR: Skipped RegionDirectiveTrivia */
            private bool _isDebug = false;

            public bool IsDebug
            {
                get
                {
                    /* TODO ERROR: Skipped IfDirectiveTrivia */
                    _isDebug = true;
                    /* TODO ERROR: Skipped EndIfDirectiveTrivia */
                    return _isDebug;
                }
            }

            private bool _isRelease = false;

            public bool IsRelease
            {
                get
                {
                    /* TODO ERROR: Skipped IfDirectiveTrivia *//* TODO ERROR: Skipped DisabledTextTrivia *//* TODO ERROR: Skipped EndIfDirectiveTrivia */
                    return _isRelease;
                }
            }

            private bool _isGasway = false;

            public bool IsGasway
            {
                get
                {
                    /* TODO ERROR: Skipped IfDirectiveTrivia *//* TODO ERROR: Skipped DisabledTextTrivia *//* TODO ERROR: Skipped EndIfDirectiveTrivia */
                    return _isGasway;
                }
            }

            private bool _isRFT = false;

            public bool IsRFT
            {
                get
                {
                    /* TODO ERROR: Skipped IfDirectiveTrivia *//* TODO ERROR: Skipped DisabledTextTrivia *//* TODO ERROR: Skipped EndIfDirectiveTrivia */
                    return _isRFT;
                }
            }

            private bool _isRFTTEST = false;

            public bool IsRFTTEST
            {
                get
                {
                    /* TODO ERROR: Skipped IfDirectiveTrivia *//* TODO ERROR: Skipped DisabledTextTrivia *//* TODO ERROR: Skipped EndIfDirectiveTrivia */
                    return _isRFTTEST;
                }
            }

            private bool _isBlueflame = false;

            public bool IsBlueflame
            {
                get
                {
                    /* TODO ERROR: Skipped IfDirectiveTrivia *//* TODO ERROR: Skipped DisabledTextTrivia *//* TODO ERROR: Skipped EndIfDirectiveTrivia */
                    return _isBlueflame;
                }
            }

            private bool _isBlueflameTest = false;

            public bool IsBlueflameTest
            {
                get
                {
                    /* TODO ERROR: Skipped IfDirectiveTrivia *//* TODO ERROR: Skipped DisabledTextTrivia *//* TODO ERROR: Skipped EndIfDirectiveTrivia */
                    return _isBlueflameTest;
                }
            }

            /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
            /* TODO ERROR: Skipped RegionDirectiveTrivia */
            private string _DatabaseServer = string.Empty;

            public string DatabaseServer
            {
                get
                {
                    return _DatabaseServer;
                }

                set
                {
                    _DatabaseServer = value;
                }
            }

            private string _DatabaseFriendlyName = string.Empty;

            public string DatabaseFriendlyName
            {
                get
                {
                    return _DatabaseFriendlyName;
                }

                set
                {
                    _DatabaseFriendlyName = value;
                }
            }

            private Enums.DataBaseName _DBName;

            public Enums.DataBaseName DBName
            {
                get
                {
                    return _DBName;
                }

                set
                {
                    _DBName = value;
                }
            }

            private string _DatabaseUsername = string.Empty;

            public string DatabaseUsername
            {
                get
                {
                    return _DatabaseUsername;
                }

                set
                {
                    _DatabaseUsername = value;
                }
            }

            private string _DatabasePassword = string.Empty;

            public string DatabasePassword
            {
                get
                {
                    return _DatabasePassword;
                }

                set
                {
                    _DatabasePassword = value;
                }
            }

            private string _DatabaseName = string.Empty;

            public string DatabaseName
            {
                get
                {
                    return _DatabaseName;
                }

                set
                {
                    _DatabaseName = value;
                }
            }

            private string _DatabasePort = string.Empty;

            public string DatabasePort
            {
                get
                {
                    return _DatabasePort;
                }

                set
                {
                    _DatabasePort = value;
                }
            }

            public string ConnectionString
            {
                get
                {
                    string str = "";
                    var switchExpr = App.TheSystem.DataBaseServerType;
                    switch (switchExpr)
                    {
                        case Enums.DatabaseSystems.MySQL:
                            {
                                str += "server=" + DatabaseServer + ";";
                                str += "user id=" + DatabaseUsername + ";";
                                if (DatabasePassword.Trim().Length > 0)
                                {
                                    str += "password=" + DatabasePassword + ";";
                                }

                                str += "database=" + DatabaseName + ";";
                                str += "port=" + DatabasePort + ";";
                                break;
                            }

                        case Enums.DatabaseSystems.Microsoft_SQL_Server:
                            {
                                if (DatabaseServer.Contains(@".\"))
                                {
                                    str = "Data Source=" + DatabaseServer + ";Initial Catalog=" + DatabaseName + ";Integrated Security=True";
                                }
                                else
                                {
                                    str = "Server=tcp:" + DatabaseServer + ";";
                                    str += "User Id=" + DatabaseUsername + ";";
                                    if (DatabasePassword.Trim().Length > 0)
                                    {
                                        str += "Password=" + DatabasePassword + ";";
                                    }

                                    str += "Database=" + DatabaseName + ";";
                                }

                                break;
                            }
                    }

                    return str;
                }
            }

            /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
            /* TODO ERROR: Skipped RegionDirectiveTrivia */

            public string SystemVersion
            {
                get
                {
                    return FileVersionInfo.GetVersionInfo(System.Reflection.Assembly.GetExecutingAssembly().Location).FileVersion;
                }
            }

            private int _SuperAdminID = 0;

            public int SuperAdminID
            {
                get
                {
                    return _SuperAdminID;
                }

                set
                {
                    _SuperAdminID = value;
                }
            }

            private string _SupportEmailFrom = string.Empty;

            public string SupportEmailFrom
            {
                get
                {
                    return _SupportEmailFrom;
                }

                set
                {
                    _SupportEmailFrom = value;
                }
            }

            private string _SupportEmailTo = string.Empty;

            public string SupportEmailTo
            {
                get
                {
                    return _SupportEmailTo;
                }

                set
                {
                    _SupportEmailTo = value;
                }
            }

            private string _SalesEmailFrom = string.Empty;

            public string SalesEmailFrom
            {
                get
                {
                    return _SalesEmailFrom;
                }

                set
                {
                    _SalesEmailFrom = value;
                }
            }

            public string InfoEmail { get; set; }

            private string _SMTPServer = string.Empty;

            public string SMTPServer
            {
                get
                {
                    return _SMTPServer;
                }

                set
                {
                    _SMTPServer = value;
                }
            }

            private string _SMTPServerUsername = string.Empty;

            public string SMTPServerUsername
            {
                get
                {
                    return _SMTPServerUsername;
                }

                set
                {
                    _SMTPServerUsername = value;
                }
            }

            private string _SMTPServerPassword = string.Empty;

            public string SMTPServerPassword
            {
                get
                {
                    return _SMTPServerPassword;
                }

                set
                {
                    _SMTPServerPassword = value;
                }
            }

            private string _DocumentsLocation = string.Empty;

            public string DocumentsLocation
            {
                get
                {
                    return _DocumentsLocation;
                }

                set
                {
                    _DocumentsLocation = value;
                }
            }

            private string _TemplateLocation = string.Empty;

            public string TemplateLocation
            {
                get
                {
                    return _TemplateLocation;
                }

                set
                {
                    _TemplateLocation = value;
                }
            }

            private string _VATNO = string.Empty;

            public string VATNO
            {
                get
                {
                    return _VATNO;
                }

                set
                {
                    _VATNO = value;
                }
            }

            public string CompanyName;
            public string CompanyAddres1;
            public string CompanyAddres2;
            public string CompanyAddres3;
            public string CompanyAddres4;
            public string CompanyAddres5;
            public string CompanyPostcode;
            public string CompanyTelephoneNumber;
            public string CompanyWebsite;
            public string CompanyDomain;
            public string CompanyFullAddress;

            /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        }
    }
}